File: 2001-3000/2874-maximum-value-of-an-ordered-triplet-ii/2874-maximum-value-of-an-ordered-triplet-ii.java
class Solution {
    public long maximumTripletValue(int[] nums) {
        long ans = 0;
        int maxVal = 0, maxDiff = 0;
        
        for (int num : nums) {
            ans = Math.max(ans, (long) maxDiff * num);
            maxDiff = Math.max(maxDiff, maxVal - num);
            maxVal = Math.max(maxVal, num);
        }
        return ans;
    }
}

File: 2001-3000/2874-maximum-value-of-an-ordered-triplet-ii/2874-maximum-value-of-an-ordered-triplet-ii.cpp
class Solution {
public:
    long long maximumTripletValue(vector<int>& nums) {
        long long  ans = 0;

        long long  maxLeft = nums[0];
        long long  maxMinus = 0;

        for(int i=1;i<nums.size();i++){
            ans = max(ans,maxMinus*nums[i]);
            
            maxMinus = max(maxMinus,maxLeft - (long long )nums[i]);

            maxLeft = max(maxLeft,(long long)nums[i]);
        }

        return ans;
    }
};